ntoskrnl/mm/ARM3/mdlsup.c

```diff
@@ -1622,29 +1622,297 @@ MmAdvanceMdl(IN PMDL Mdl,
 }
 
 /*
- * @unimplemented
+ * @implemented
  */
 PVOID
 NTAPI
 MmMapLockedPagesWithReservedMapping(IN PVOID MappingAddress,
                                     IN ULONG PoolTag,
-                                    IN PMDL MemoryDescriptorList,
+                                    IN PMDL Mdl,
                                     IN MEMORY_CACHING_TYPE CacheType)
 {
-    UNIMPLEMENTED;
-    return 0;
+    PPFN_NUMBER MdlPages, LastPage;
+    PFN_COUNT PageCount;
+    BOOLEAN IsIoMapping;
+    MI_PFN_CACHE_ATTRIBUTE CacheAttribute;
+    PMMPTE PointerPte;
+    MMPTE TempPte;
+
+    ASSERT(Mdl->ByteCount != 0);
+
+    //
+    // Get the list of pages and count
+    //
+    MdlPages = MmGetMdlPfnArray(Mdl);
+    PageCount = ADDRESS_AND_SIZE_TO_SPAN_PAGES(MmGetMdlVirtualAddress(Mdl),
+                                               Mdl->ByteCount);
+    LastPage = MdlPages + PageCount;
+
+    //
+    // Sanity checks
+    //
+    ASSERT((Mdl->MdlFlags & (MDL_MAPPED_TO_SYSTEM_VA |
+                             MDL_SOURCE_IS_NONPAGED_POOL |
+                             MDL_PARTIAL_HAS_BEEN_MAPPED)) == 0);
+    ASSERT((Mdl->MdlFlags & (MDL_PAGES_LOCKED | MDL_PARTIAL)) != 0);
+
+    //
+    // Get the correct cache type
+    //
+    IsIoMapping = (Mdl->MdlFlags & MDL_IO_SPACE) != 0;
+    CacheAttribute = MiPlatformCacheAttributes[IsIoMapping][CacheType];
+
+    //
+    // Get the first PTE we reserved
+    //
+    ASSERT(MappingAddress);
+    PointerPte = MiAddressToPte(MappingAddress) - 2;
+    ASSERT(!PointerPte[0].u.Hard.Valid &&
+           !PointerPte[1].u.Hard.Valid);
+
+    //
+    // Verify that the pool tag matches
+    //
+    TempPte.u.Long = PoolTag;
+    TempPte.u.Hard.Valid = 0;
+    if (PointerPte[1].u.Long != TempPte.u.Long)
+    {
+        KeBugCheckEx(SYSTEM_PTE_MISUSE,
+                     0x104, /* Trying to map an address it does not own */
```

Comment:
We don't have defines for these values yet?

Reply:
I was not able to find any defines for these error values including 0x104, 0x105, 0x108, and 0x109, but I suspect they must be out on the Internet somewhere.

Reply:
@oleg-dubinskiy could help because it's him who added the other similar ones used elsewhere for this purpose.

Reply:
@Doug-Lyons @HBelusca I recently added some defines here: https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/mm/ARM3/miarm.h;h=7515b4c506b0cd571fce20c7f0da89673acd9754;hb=HEAD#l157 with my previous PR. You can use them as well here if needed/add some more ones for other bugcheck subcodes too. :wink:

Reply:
These defines are internal, so they are not documented in any way in the web, I added the names for them randomly, from my head, since Jerome suggested me to do that in the first oldest PR. But the subcodes itself should be definitely Windows compatible. Defines are just an aliases for them.

Reply:
I have added defines for these four values of 0x104, 0x105, 0x108, and 0x109 in miarm.h. Thanks.